### PR TITLE
ENH: tags stashed on env

### PR DESCRIPTION
**Pull request by story645**
Pulled out of #84 for easier review, this PR:
* moves the tag list for each doc to the `app.env.metadata[docname]['tags']` dictionary, which makes the list global
* patches Path with copytree so that pathlib can be used in conftest

Questions:
* parsing the file in the directive because myst parser returns a blank `block_text`, my preference would be to just use the argument, but then it won't preserve the full white space -> is it very important that ``tag    4`` get rendered as ``tag    4`` and not ``tag 4`` -> this is now the 2nd commit for comparison
* not sure what the "WARNING: 'any' reference target not found" error means, given the badges still seem to kinda work?

### CODE REVIEW

**Comment by JWCook** [comment]
> my preference would be to just use the argument, but then it won't preserve the full white space

I agree, I don't think that's very important since Markdown already collapses consecutive spaces.

> not sure what the "WARNING: 'any' reference target not found" error mean

I know I've seen that before. It might be something along the lines of "something was interpreted as a `:ref:` that wasn't intended to be a `:ref:`." It could be either a minor syntax problem or something that could be ignored. Are you seeing this in the tests, or the doc build for sphinx-tags itself?

Either way, not a very helpful error message. I'd rate it 2/10.

**Comment by melissawm** [comment]
I'm not sure what's happening. Locally, I am seeing the "any" error on the regular docs build, and the tags are collected but the links are not actually generated. I'll try debugging some more if I find the time today or tomorrow but I'm going to be out all next week on vacation. If you folks find a fix please just merge and move forward with it, the general idea is good on my side 😄

File: src/sphinx_tags/__init__.py [hunk 1]
```diff
@@ -294,24 +281,18 @@ def assign_entries(app):
...
-    # Get document paths in the project that match specified file extensions
-    doc_paths = get_matching_files(
-        app.srcdir,
-        include_patterns=[f"**.{extension}" for extension in app.config.tags_extension],
-        exclude_patterns=app.config.exclude_patterns,
-    )
-
-    for path in doc_paths:
-        entry = Entry(Path(app.srcdir) / path)
+    for docname in app.env.found_docs:
```
Comment (JWCook) [inline src/sphinx_tags/__init__.py:284]: I had no idea this was a thing! Good find. I was looking for something like that for #58 but somehow missed it.

File: src/sphinx_tags/__init__.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 """
 import os
 import re
+from collections import defaultdict
```
Comment (JWCook) [inline src/sphinx_tags/__init__.py:6]: Unused imports: this and `get_matching_files`. Those could be removed in your other PR, though.

**Review by JWCook — COMMENTED** [review]
These look like some nice improvements. Thanks for putting this together and fighting through all those testing issues!

**Review by JWCook — APPROVED** [review]

**Comment by JWCook** [comment]
@story645 Okay, I think I see what's going on. Those errors (`File Not Found: _tags/*.html` and `WARNING: 'any' reference target not found`) were legit, and tag pages aren't getting generated in normal builds. For example:
```sh
rm -rf docs/{_tags,_build}
make -C docs html
ls docs/_tags
```

At the point where we're checking `app.env.metadata["tags"]` in `assign_entries()`, I think the `builder-inited` event hasn't yet run, so `TagLinks.run()` hasn't been called yet, so that metadata is empty. It looks like this will only be populated if the build is run twice within a single `SphinxApp` instance, which is why `test_tag_pages()` works after `test_build()`, but not on its own.

One possible fix may be to connect `update_tags()` to a different event later in the build process ([reference](https://www.sphinx-doc.org/en/master/extdev/appapi.html#sphinx-core-events)).

**Comment by JWCook** [comment]
I also looked into this a bit and haven't quite figured it out yet, but I do feel somewhat closer to understanding how the build events work. Honestly I wish Sphinx's docs on the topic were better!

We probably wouldn't need to revert all of your changes, but maybe we could add the parsing back to the `Event` class if we can't figure it out by the end of the week?

**Comment by JWCook** [comment]
> Either that or a non env global variable maybe?

The order of operations is still the tricky part there: we need `TagLinks` to run first, then `update_tags()`, and all before `toctree` runs so it can pick up the pages generated by `update_tags()`. Using the `source-read` event (run for each individual document) with a higher priority than `toctree` might work... except for the tag overview page since it needs access to all tags.

> Another option is refactoring all the things into the kinds domain/index/directive model here

That looks promising! So would that look something like this?
* 1 index class for tag pages
* 1 index class for the tag overview page
* A domain class containing those indices and the tags directive

**Pull request merged**